File: Excel/643 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/643 Shift Numbers by One Position.xlsx"
input = read_excel(path, range = "A1:A8")
test  = read_excel(path, range = "B1:B8")

shift_digits <- function(s) {
  split <- str_split(s, "(?<=\\d)(?=\\D)|(?<=\\D)(?=\\d)", simplify = TRUE)
  digit_idxs <- which(str_detect(split, "\\d"))
  
  if (length(digit_idxs) > 0) {  
    shifted_digits <- split[digit_idxs] %>% 
      {c(tail(., 1), head(., -1))}
    split[digit_idxs] <- shifted_digits
  }
  str_c(split, collapse = "")
}

  
result = input %>%
  mutate(answer = map_chr(Strings, shift_digits))

all.equal(result$answer, test$`Answer Expected`)
#> [1] TRUE

File: Excel/643 Challenge.py
import pandas as pd
import re

path = "643 Shift Numbers by One Position.xlsx"
input = pd.read_excel(path, usecols="A", nrows=8)
test = pd.read_excel(path, usecols="B", nrows=8)

def shift_digits(s):
    split = re.split(r'(\d+|\D+)', s)
    split = [x for x in split if x]
    digit_idxs = [i for i, part in enumerate(split) if part.isdigit()]
    
    if digit_idxs:
        shifted_digits = [split[digit_idxs[-1]]] + [split[i] for i in digit_idxs[:-1]]
        for i, idx in enumerate(digit_idxs):
            split[idx] = shifted_digits[i]
    
    return ''.join(split)

input['answer'] = input.iloc[:, 0].apply(shift_digits)

print(input['answer'].equals(test['Answer Expected']))  #True
